File: sdk/python/example/example_ws.py
import logging
import os
import time

from kucoin_universal_sdk.api import DefaultClient
from kucoin_universal_sdk.generate.futures.futures_public import FuturesPublicWS
from kucoin_universal_sdk.generate.futures.futures_public import TickerV2Event
from kucoin_universal_sdk.generate.spot.spot_public import SpotPublicWS
from kucoin_universal_sdk.generate.spot.spot_public import TickerEvent
from kucoin_universal_sdk.model import ClientOptionBuilder
from kucoin_universal_sdk.model import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model import WebSocketClientOptionBuilder


def ws_example():
    logging.basicConfig(
        level=logging.INFO,
        format='%(asctime)s %(levelname)s - %(message)s',
        datefmt='%Y-%m-%d %H:%M:%S'
    )

    # Retrieve API secret information from environment variables
    key = os.getenv("API_KEY", "")
    secret = os.getenv("API_SECRET", "")
    passphrase = os.getenv("API_PASSPHRASE", "")

    # Set specific options, others will fall back to default values
    ws_client_option = (
        WebSocketClientOptionBuilder().build()
    )

    # Create a client using the specified options
    client_option = (
        ClientOptionBuilder()
        .set_key(key)
        .set_secret(secret)
        .set_passphrase(passphrase)
        .set_websocket_client_option(ws_client_option)
        .set_spot_endpoint(GLOBAL_API_ENDPOINT)
        .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
        .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
        .build()
    )
    client = DefaultClient(client_option)
    # Get the websocket service
    kucoin_ws_service = client.ws_service()
    spot_ws_example(kucoin_ws_service.new_spot_public_ws())
    futures_ws_example(kucoin_ws_service.new_futures_public_ws())


def spot_ws_example(spot_public_ws: SpotPublicWS):
    try:
        # Start WebSocket
        spot_public_ws.start()
    except Exception as e:
        logging.error(f"failed to start spot public websocket: {e}")
        return

    try:
        def ticker_event_callback(topic: str, subject: str, data: TickerEvent) -> None:
            # Process logic
            logging.info(
                f"received ticker event {topic} {subject} {data.sequence} {data.price} {data.time} {data.size}")

        # Get ticker
        try:
            sub_id = spot_public_ws.ticker(["BTC-USDT"], ticker_event_callback)
        except Exception as e:
            logging.fatal(f"subscribe error: {e}")
            return

        # Triggered when certain conditions are met
        time.sleep(10)

        # Unsubscribe by sub id
        try:
            spot_public_ws.unsubscribe(sub_id)
        except Exception as e:
            logging.fatal(f"unsubscribe error: {e}, id: {sub_id}")
    finally:
        spot_public_ws.stop()


def futures_ws_example(futures_public_ws: FuturesPublicWS):
    try:
        # Start WebSocket
        futures_public_ws.start()
    except Exception as e:
        logging.fatal(f"failed to start futures public websocket: {e}")
        return

    try:
        def ticker_event_v2_callback(topic: str, subject: str, data: TickerV2Event) -> None:
            logging.info(f"received ticker event {data.to_json()}")

        # Get Ticker
        try:
            sub_id = futures_public_ws.ticker_v2("XBTUSDTM", ticker_event_v2_callback)
        except Exception as e:
            logging.fatal(f"subscribe error: {e}")
            return
        # Triggered when certain conditions are met
        time.sleep(10)

        # Unsubscribe
        try:
            futures_public_ws.unsubscribe(sub_id)
        except Exception as e:
            logging.fatal(f"unsubscribe error: {e}, id: {sub_id}")
    finally:
        futures_public_ws.stop()


if __name__ == "__main__":
    ws_example()

File: sdk/php/example/ExampleWs.php
<?php

use KuCoin\UniversalSDK\Api\DefaultClient;
use KuCoin\UniversalSDK\Common\Logger;
use KuCoin\UniversalSDK\Generate\Spot\SpotPublic\AllTickersEvent;
use KuCoin\UniversalSDK\Model\ClientOptionBuilder;
use KuCoin\UniversalSDK\Model\Constants;
use KuCoin\UniversalSDK\Model\TransportOptionBuilder;
use KuCoin\UniversalSDK\Model\WebSocketClientOptionBuilder;
use React\EventLoop\Loop;

include '../vendor/autoload.php';

function wsExample()
{
    // Credentials & setup
    $key = getenv('API_KEY') ?: '';
    $secret = getenv('API_SECRET') ?: '';
    $passphrase = getenv('API_PASSPHRASE') ?: '';

    $brokerName = getenv('BROKER_NAME');
    $brokerKey = getenv('BROKER_KEY');
    $brokerPartner = getenv('BROKER_PARTNER');

    $httpTransportOption = (new TransportOptionBuilder())
        ->setKeepAlive(true)
        ->setMaxConnections(10)
        ->build();

    $websocketTransportOption = (new WebSocketClientOptionBuilder())->build();

    $clientOption = (new ClientOptionBuilder())
        ->setKey($key)
        ->setSecret($secret)
        ->setPassphrase($passphrase)
        ->setBrokerName($brokerName)
        ->setBrokerKey($brokerKey)
        ->setBrokerPartner($brokerPartner)
        ->setSpotEndpoint(Constants::GLOBAL_API_ENDPOINT)
        ->setFuturesEndpoint(Constants::GLOBAL_FUTURES_API_ENDPOINT)
        ->setBrokerEndpoint(Constants::GLOBAL_BROKER_API_ENDPOINT)
        ->setTransportOption($httpTransportOption)
        ->setWebSocketClientOption($websocketTransportOption)
        ->build();

    // Create or get the global event loop
    $loop = Loop::get();

    $client = new DefaultClient($clientOption, $loop);
    $spotWs = $client->wsService()->newSpotPublicWS();

    // Start connection
    $spotWs->start()->then(function () use ($spotWs, $loop) {
        Logger::info("WebSocket started");

        // Subscribe to allTickers
        return $spotWs->allTickers(
        // Called when data is received
            function (string $topic, string $subject, AllTickersEvent $data) {
                Logger::info("Ticker update", [
                    'topic' => $topic,
                    'subject' => $subject,
                    'bestBid' => $data->bestBid,
                    'bestAsk' => $data->bestAsk,
                ]);
            },
            // Called when subscription is successful
            function (string $id) use ($spotWs, $loop) {
                Logger::info("Subscribed with ID: $id");

                // Schedule unsubscribe and shutdown after 5 seconds
                $loop->addTimer(5, function () use ($id, $spotWs) {
                    Logger::info("Unsubscribing...");
                    $spotWs->unSubscribe($id)->finally(function () use ($spotWs) {
                        $spotWs->stop();
                    });
                });
            },
            // Called when subscription fails
            function (Exception $e) use ($spotWs) {
                Logger::error("Subscription failed", ['error' => $e->getMessage()]);
                $spotWs->stop();
            }
        );
    })->catch(function (Exception $e) use ($spotWs) {
        Logger::error("Failed to start", ['error' => $e->getMessage()]);
        $spotWs->stop();
    });

    // Run the event loop to process async tasks
    $loop->run();
}

if (php_sapi_name() === 'cli') {
    wsExample();
}
